pkg/bootstrap/platform/gcp.go

```diff
@@ -249,33 +253,64 @@ func waitForMetadataSuppliers(suppliers []metadataSupplier, md map[string]string
 	return &wg
 }
 
+// Zones are in the form <country_code>-<region_suffix>-<zone_suffix>.
+var zoneRE = regexp.MustCompile(`^([^-]+-[^-]+)-[^-]+$`)
+
 // Converts a GCP zone into a region.
 func zoneToRegion(z string) (string, error) {
-	// Zones are in the form <region>-<zone_suffix>, so capture everything but the suffix.
-	re := regexp.MustCompile("(.*)-.*")
-	m := re.FindStringSubmatch(z)
+	m := zoneRE.FindStringSubmatch(z)
 	if len(m) != 2 {
 		return "", fmt.Errorf("unable to extract region from GCP zone: %s", z)
 	}
 	return m[1], nil
 }
 
+// Returns the zone where the pod is located in.
+func (e *gcpEnv) getPodZone() (string, error) {
+	z := e.cachedZone.Load()
+	if z != "" {
+		return z, nil
+	}
+	ctx, cfn := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cfn()
+	z, err := zoneFn(ctx)
+	if err != nil {
+		return "", err
+	}
+	e.cachedZone.Store(z)
+	return z, nil
+}
+
+// Return the zone guessed from the cluster location.
+func zoneFromClusterLocation(loc string) string {
+	if zoneRE.MatchString(loc) {
+		return loc
+	}
+	// Fallback to "-unknown" zone suffix if the given location is regional.
+	return loc + "-unknown"
+}
+
 // Locality returns the GCP-specific region and zone.
 func (e *gcpEnv) Locality() *core.Locality {
 	var l core.Locality
-	loc := e.Metadata()[GCPLocation]
-	if loc == "" {
-		log.Warnf("Error fetching GCP zone: %v", loc)
-		return &l
+	z, err := e.getPodZone()
```

Comment:
In the old code, we were very careful to allow getting all required information without hitting the metadata server if `GCPStaticMetadata` is set. This breaks this. You might care about preserving that behavior

Reply:
yeah. you're right.
Hmm, OK, let's put away this PR for a while. 
I will check if there is good way to get the Pod's zone without the metadata server. 
Then, I will get back to you.

Thanks!

Reply:
There is no way to avoid a callout here I think to get the node zone since the static metadata can only have cluster metadata.

Reply:
@howardjohn We researched more and concluded that MDS is only the viable way to get the correct zone for now. 

Probably, if https://github.com/kubernetes/enhancements/pull/4747 is realized, we can use the node annotation for getting the zone.